### ISSUE TITLE
Easier way to install/update/add plugins or themes

### ISSUE THREAD

**Issue by naplaya** [bludit/bludit#994]
Feature Request:
I think there is a simple solution to install plugins.

It could be an extra UI, where you can simply instal plugins from the homepage without having to manually download them from the homepage, unzip them and then move them into the plugins folder and activate them, of course.

This UI could look like this and be in the plugins UI.
<img width="503" alt="aaaaa" src="https://user-images.githubusercontent.com/24904919/54480175-7b3b8980-4825-11e9-93be-a1a990391e4d.PNG">
<img width="523" alt="dddddddd" src="https://user-images.githubusercontent.com/24904919/54480173-74147b80-4825-11e9-933a-3ce0361d9822.PNG">

The User could also upload a plugin from his computer to bludit CMS

That could work just as well for themes.

**Comment by philippd1** [comment]
great idea!👌
I personally think, the best way of quickly realizing this actually would be to build a plugin to load other plugins:
my basic idea:
- choose plugin
- download plugin ZIP file from the GitHub repo
- unzip to plugin folder
- delete ZIP
- activate plugin

actually, this should be a pretty simple task.😁
I'll definitely look into this in the very near future😉

**Comment by philippd1** [comment]
Okay, so this is my progress so far:
I started to integrate the installer into the admin sidebar for ease of use
![image](https://user-images.githubusercontent.com/17651032/54564430-14a4a000-49cc-11e9-8240-0a497ffd8a29.png)
checking if curl is installed, if not: display alert for now

![image](https://user-images.githubusercontent.com/17651032/54564519-4a498900-49cc-11e9-98cb-19f68b5b58e2.png)
if cURL is enabled, display the form (method="GET"; action="") with a url input and a submit button
When clicking on the button, the form submits and now comes the tricky part where I'm currently stuck.
![image](https://user-images.githubusercontent.com/17651032/54564709-c17f1d00-49cc-11e9-8fa2-a4371d972f02.png)

I need some way to download the ZIP into the directory HTML_PATH_PLUGINS (defined in init.php)
the simples approach I took was this:
```
file_put_contents(HTML_PATH_PLUGINS."plugin-download.zip", fopen("https://github.com/philippd1/bludit-auto-update/archive/master.zip", 'r'));
```
This does not work tho.

Maybe @SamBrishes or @dignajar know how to help here?

all of the current code can be found here: https://pastebin.com/9203Ykqf

**Comment by naplaya** [comment]
damn i just finished with the ui of the plugin. just download, extract and  activate is missing :/

**Comment by philippd1** [comment]
hahah😂
well if we get the zip download to work it's gonna be easy

**Comment by naplaya** [comment]
i got an idea from stack overflow
but no time ti test. its late (in europe)
```
$f = file_put_contents("my-zip.zip", fopen("http://www.externalsite.com/zipfile.zip", 'r'), LOCK_EX);
if(FALSE === $f)
  die("Couldn't write to file.");
        
  $zip = new ZipArchive;
  $res = $zip->open('my-zip.zip');
  if ($res === TRUE) {
    $zip->extractTo('/extract-here');
    $zip->close();
          //
   } else {
          //
  }
```

**Comment by philippd1** [comment]
will quickly try this, then gonna hit the hay too ^^ (I'm from Germany 😉)

**Comment by naplaya** [comment]
me too xD

attached my progress, looks quite like plugin settings

![image](https://user-images.githubusercontent.com/24904919/54568107-52a6c180-49d6-11e9-98b2-8ce8c38a0998.png)

**Comment by philippd1** [comment]
dang, that looks clean👌
when trying out your code I'm getting 'Couldn't write to file.'
I think I'll have to look into this tomorrow, as it also seems that I cannot even write a simple file like this
```
$myfile = fopen("test.txt", "w") or die("Unable to open file!");
fwrite($myfile, "test msg here");
fclose($myfile);
```

**Comment by naplaya** [comment]
yes youre right let me just setup a repo. as you suggested ealier, i build a plugin.
[https://github.com/naplaya/plugin-downlaod](https://github.com/naplaya/plugin-downlaod)

**Comment by naplaya** [comment]
thought there was an auto update plugin and found this. may it helps
[https://github.com/Hatschipuh/plugin_update](https://github.com/Hatschipuh/plugin_update)

**Comment by naplaya** [comment]
it seems i could downlaod zip and extract it to PATH_PLUGINS, but dont know if it works on other systems

**Comment by philippd1** [comment]
Sounds great!
I will either check it later today or tomorrow 😉

**Comment by naplaya** [comment]
iam done for today, download and extract works fine but plugins were not activated. i gues there is an error finding the name of the class for activation.

**Comment by Hatschipuh** [comment]
naplaya are you german?

**Comment by naplaya** [comment]
Yes, I am

Hatschipuh <notifications@github.com> schrieb am Mi., 20. März 2019, 12:49:

**Comment by SamBrishes** [comment]
Hellow,

Another German here... Just Kidding, I'm not German, just living in Germany. The idea is good, I'll check the plugin and test it on my both environments :D.

Sincerely,
Sam.

**Comment by anaggh** [comment]
I think Bludit should include a webhook on Github commit that builds/generates a master-plugins.json of all plugins (without screenshots for faster loads).

Then from the Admin Interface, this master-plugins.json could be called via AJAX to list available plugins.

On install, it could look for `download_url` of plugin, use the inbuilt `TCP::` class, unzip it. Search the directory recursively for the `metadata.json` path. Consider that as the root of plugin and then move to plugins path. We could also include plugin update alerts by comparing installed version and `version` from the master-plugins.json

Same thing could be for themes, should probably be a part of core bludit like Bludit Package Manager :D

**Comment by philippd1** [comment]
> Same thing could be for themes, should probably be a part of core bludit like Bludit Package Manager :D

agreed 👍
I will work on a theme installer soon ;)

**Comment by philippd1** [comment]
Here's the theme installer plugin: https://github.com/philippd1/theme-download

working with AJAX JS so it nearly instantly loads the list of available themes
if @dignajar accepts this, it should soon be available ([PR](https://github.com/bludit/plugins-repository/pull/60))

**Comment by misterunknown** [comment]
Any updates on this? Could this be extended to be able to update bludit itself via the frontend?

**Comment by philippd1** [comment]
> Any updates on this? Could this be extended to be able to update bludit itself via the frontend?
@misterunknown yes, take a look at https://github.com/bludit/bludit/issues/1057#issuecomment-586924593 + https://github.com/philippd1/bludit-auto-update 😎

**Comment by misterunknown** [comment]
Thanks for the link, nice to see the progress.

**Comment by dignajar** [comment]
Hello, with the new version of Bludit v4.0 rc1, I would like to close the old Github issues. If you feel that your issue is not resolved in the latest version, create a new ticket.

- Help and Support use the Forum https://forum.bludit.org
- Bugs and new requests here in Github https://github.com/bludit/bludit/issues
